pkg/sql/catalog/typedesc/type_desc.go

```diff
@@ -679,6 +688,26 @@ func (desc *immutable) ValidateForwardReferences(
 			vea.Report(errors.AssertionFailedf("aliased type %q (%d) is dropped", typ.GetName(), typ.GetID()))
 		}
 	}
+
+	if desc.GetKind() == descpb.TypeDescriptor_COMPOSITE {
+		for _, t := range desc.Composite.TupleContents {
+			if !t.UserDefined() {
+				continue
+			}
+			id, err := GetUserDefinedTypeDescID(t)
+			if err != nil {
+				vea.Report(err)
+				continue
+			}
+			typeDesc, err := vdg.GetTypeDescriptor(id)
+			if err != nil {
+				vea.Report(err)
+			} else if typeDesc.Dropped() {
+				vea.Report(errors.AssertionFailedf("referenced type %q (%d) is dropped",
+					typeDesc.GetName(), typeDesc.GetID()))
+			}
+		}
```

Comment:
Is this code exercised at all now? Shouldn't we simply be erroring when `t.UserDefined()` is true?